usr/sbin/p11sak/p11sak.c

```diff
@@ -999,34 +1114,49 @@ CK_BBOOL is_mech_supported(CK_SLOT_ID slot, CK_MECHANISM *pmech,
  */
 static CK_RV tok_key_gen(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
                          CK_ULONG keylength, CK_MECHANISM *pmech,
-                         char *attr_string, CK_OBJECT_HANDLE *phkey,
-                         char *label)
+                         char *attr_string, char *custom_attr_string, 
+                         CK_OBJECT_HANDLE *phkey, char *label)
 {
     CK_RV rc;
-    int i = 0;
+    int j, i = 0;
 
     /* Boolean attributes (cannot be specified by user) */
     CK_BBOOL a_token = ckb_true; // always true
 
     /* Non-boolean attributes */
     CK_ULONG a_value_len = keylength / 8;
 
-    /* Standard template */
-    CK_ATTRIBUTE key_attr[3 + KEY_MAX_BOOL_ATTR_COUNT] = {
+    /**Standard template 
+     * 128 is the current count of attributes defined in pkcs11types.h
+     * last updated 08.01.2022
+     */
+    CK_ATTRIBUTE key_attr[128] = {
         { CKA_TOKEN, &a_token, sizeof(CK_BBOOL) }, 
         { CKA_VALUE_LEN, &a_value_len, sizeof(CK_ULONG) }, 
         { CKA_LABEL, label, strlen(label) } 
     };
     CK_ULONG num_attrs = 3;
+    j = (int) num_attrs;
     
     /* set boolean attributes, set template from string 
     attr_string length is checked in parse_gen_key_args to avoid memory problems */
     if (attr_string) {
-        for (i = 0; i < (int) strlen(attr_string); i++) {
+        for (i = 0; i < (int) strlen(attr_string) && i < (int)(sizeof(key_attr) / sizeof(CK_ATTRIBUTE)); i++) {
```

Comment:
That is an off-by-3.  `i` starts at 0 and can run to max of 128.  So we store attributes 3 to 131 since we add `num_attrs` to `i`.

Reply:
Done